### DISCUSSION TITLE
How to access the structure of type "const sdbus::Struct<std::string, std::string, std::string, std::string>& DbusInfo"

### DISCUSSION THREAD

**Discussion by Siddharthav1** [Kistler-Group/sdbus-cpp#292] [Q&A]
Hello,

   Recently I started working on sdbus. I have one xml where on of the methods take structure as its agruement. But on generating the code from it I got the structure in the below syntax. i.e 

`"const sdbus::Struct<std::string, std::string, std::string, std::string>& DbusInfo".`

Here I need to store and print the values of the member variables of the structure, but the nme of the variables are not coming. May I know if I have missed something while creating xml or is there any different way to access this type of structure please.

Regards,
Siddhartha V

**Comment by sangelovic** [comment]
Hello, there is no support for names of D-Bus structure members in D-Bus IDL. So what you have in your IDL XML as argument signature is probably this: `<arg type="(ssss)" name="DbusInfo" direction="in" />`. As you can see, you have just the name of the argument here, not of individual struct members. What you get out of this is what you mentioned in your post. Now, `sdbus::Struct` is implemented on top of [`std::tuple`](https://en.cppreference.com/w/cpp/utility/tuple), so you can use its API to get access to the individual members of the `sdbus::Struct` (a-ka `std::tuple`).

I plan to implement support for custom user structures, like mentioned in [this discussion](https://github.com/Kistler-Group/sdbus-cpp/discussions/281), at the convenience API level. At the XML IDL level (generated API code level), we would need to extend the XML format so you can define structures there, see #275. If you wish to contribute here, you are welcome :)
